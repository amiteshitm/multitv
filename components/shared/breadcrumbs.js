import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import Link from 'next/link'

import { makeStyles } from '@material-ui/core/styles'
import Icon from '@material-ui/core/Icon'
import Breadcrumbs from '@material-ui/core/Breadcrumbs'
import Paper from '@material-ui/core/Paper'
import grey from '@material-ui/core/colors/grey'

import globalStyle from 'shared/styles'

const useStyles = makeStyles(theme => ({
  root: {
    padding: theme.spacing(2),
    margin: theme.spacing(2),
    background: grey[100],
  },
  homeIcon: {
    marginRight: theme.spacing(1),
  },
  bcSeparator: {
    padding: `0 ${theme.spacing(1)}px 0`,
  },
  li: {
    display: 'flex',
  },
  layout: globalStyle(theme).layout,
}))

const BreadcrumbBar = ({ path }) => {
  const classes = useStyles()

  const breadcrumbs = path.map((p, idx) => {
    const lastIndex = idx === path.length - 1

    if (p.href === '/') {
      return (
        <Fragment key={idx}>
          <Icon className={classes.homeIcon}>home</Icon>
          <Link href="/" as="/">
            <a>Home</a>
          </Link>
        </Fragment>
      )
    } else {
      return (
        <Fragment key={idx}>
          {!lastIndex && (
            <Link href={p.href} as={p.as}>
              <a>{p.name}</a>
            </Link>
          )}
          {lastIndex && <span>{p.name}</span>}
        </Fragment>
      )
    }
  })

  return (
    <Paper elevation={0} className={classnames(classes.root, classes.layout)}>
      <Breadcrumbs classes={{ li: classes.li }} aria-label="breadcrumb">
        {breadcrumbs}
      </Breadcrumbs>
    </Paper>
  )
}

BreadcrumbBar.propTypes = {
  path: PropTypes.array.isRequired,
}

export default BreadcrumbBar
