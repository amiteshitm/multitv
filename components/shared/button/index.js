import React from 'react'
import PropTypes from 'prop-types'
import { Button } from 'antd'

export default function WrappedButton({ children, ...props }) {
  return <Button {...props}>{children}</Button>
}

WrappedButton.propTypes = {
  children: PropTypes.node.isRequired
}
