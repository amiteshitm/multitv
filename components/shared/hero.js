import React from 'react'
import PropTypes from 'prop-types'

import classNames from 'classnames'

import { makeStyles } from '@material-ui/core/styles'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'

import globalStyles from 'shared/styles'

const useStyles = makeStyles(theme => ({
  root: {
    flexGrow: 1,
  },
  paper: {
    padding: theme.spacing(2),
    margin: '16px auto',
    width: '100%',
    borderBottom: '1px solid #ccc',
  },
  rightActions: {
    display: 'flex',
    alignItems: 'center',
  },
  layout: globalStyles(theme).layout,
}))

const Hero = ({ name, header }) => {
  const classes = useStyles()

  return (
    <div className={classNames(classes.root, classes.layout)}>
      <div className={classes.paper}>
        <Grid container spacing={16}>
          <Grid item xs={12} sm container>
            <Grid item xs container direction="column" spacing={16}>
              <Grid item xs>
                <Typography gutterBottom component="h1" variant="h4">
                  {name}
                </Typography>
                <Typography component="h2" variant="h6" gutterBottom>
                  {header}
                </Typography>
              </Grid>
            </Grid>
          </Grid>
        </Grid>
      </div>
    </div>
  )
}

Hero.propTypes = {
  name: PropTypes.string.isRequired,
  header: PropTypes.string,
}

export default Hero
