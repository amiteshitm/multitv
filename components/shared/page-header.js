import React from 'react'
import PropTypes from 'prop-types'

import { makeStyles } from '@material-ui/core/styles'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'

const useStyles = makeStyles(theme => ({
  container: {
    padding: theme.spacing(2),
    margin: '16px auto',
    justifyContent: 'center',
    textAlign: 'center',
    borderBottom: '1px solid #ccc',
    width: 'auto',
  },
}))

const PageHeader = ({ title, subTitle }) => {
  const classes = useStyles()

  return (
    <Grid className={classes.container} container>
      <Grid item>
        <Typography gutterBottom component="h1" variant="h4">
          {title}
        </Typography>
        <Typography gutterBottom>{subTitle}</Typography>
      </Grid>
    </Grid>
  )
}

PageHeader.propTypes = {
  title: PropTypes.string.isRequired,
  subTitle: PropTypes.string,
}

export default PageHeader
