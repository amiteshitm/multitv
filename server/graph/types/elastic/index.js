import SoftwareES from './software'

export default SoftwareES
