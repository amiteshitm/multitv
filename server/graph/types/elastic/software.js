import elasticsearch from 'elasticsearch'
import { composeWithElastic } from 'graphql-compose-elasticsearch'

import mapping from 'shared/elastic/mapping'

require('dotenv').config()

const SoftwareElasticTC = composeWithElastic({
  graphqlTypeName: 'SoftwareES',
  elasticIndex:
    process.env.NODE_ENV === 'test'
      ? 'softwares_test'
      : process.env.ELASTIC_INDEX_ALIAS,
  elasticType: 'software',
  elasticMapping: mapping,
  elasticClient: new elasticsearch.Client({
    host: `${process.env.ELASTIC_HOST}`,
    apiVersion: '7.3',
    log: 'trace'
  }),
  // elastic mapping does not contain information about is fields are arrays or not
  // so provide this information explicitly for obtaining correct types in GraphQL
  pluralFields: ['categories', 'uploads', 'logo', 'tags']
})

export default {
  softwareES: SoftwareElasticTC.getResolver('search').getFieldConfig(),
  softwarePagination: SoftwareElasticTC.getResolver(
    'searchPagination'
  ).getFieldConfig(),
  softwareConnection: SoftwareElasticTC.getResolver(
    'searchConnection'
  ).getFieldConfig()
}
