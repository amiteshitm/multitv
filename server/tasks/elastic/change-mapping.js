const { Client } = require('elasticsearch')
const mapping = require('../../../shared/elastic/mapping').default

const client = new Client({ node: `${process.env.ELASTIC_HOST}` })

// To run the tasks first build using: webpack --mode development --config webpack.tasks.config.js
// or yarn run build-tasks in the root directory of project.
// Then run, jake -f dist/main.js esChangeMapping.

desc('The task updates the mapping with the latest mapping')
task('esChangeMapping', { async: true }, async function (params) {
  try {
    const oldIndexName = 'softwares'
    const newIndexName = 'softwares_v2'
    const aliasName = 'softwares_alias'

    // create alias.
    // Next time make softwares as alias and just keep changing the versions
    // of index with alias pointing to the version numbered index
    // e.g alias: softwares, index: softwares_v2
    await client.indices.putAlias({
      index: 'softwares',
      name: aliasName,
    })

    // step 2: create new index with new mapping.
    await client.indices.create({
      index: newIndexName,
    })

    let result = await client.indices.putMapping({
      index: newIndexName,
      type: 'software',
      include_type_name: true,
      body: mapping,
    })

    console.log('After putMapping')
    console.log(result)

    // Step: 3 Reindex into the new index
    result = await client.reindex({
      body: {
        source: {
          index: oldIndexName,
        },
        dest: {
          index: newIndexName,
        },
      },
    })

    console.log('After reindexing')
    console.log(result)

    // Step 4: Point the alias to the new index now.
    result = await client.indices.updateAliases({
      body: {
        actions: [
          { remove: { index: oldIndexName, alias: aliasName } },
          { add: { index: newIndexName, alias: aliasName } },
        ],
      },
    })

    console.log('After update aliases')
    console.log(result)
  } catch (err) {
    console.log(err.message)
  } finally {
    this.complete()
  }
})

jake.addListener('complete', function () {
  setImmediate(function () {
    process.exit()
  })
})
