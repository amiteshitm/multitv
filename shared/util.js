import React from 'react'
import PropTypes from 'prop-types'

import capitalize from 'lodash/capitalize'

import { Button } from 'antd'
import Link from 'next/link'

const NextButtonLink = itemProps => (
  <Link href={itemProps.href}>{itemProps.children}</Link>
)

const isParent = feature =>
  feature.parent === null || feature.parent === undefined
const isChild = feature => !isParent(feature)

const NextLink = ({ href, children, as }) => (
  <Link href={href} as={as}>
    {children}
  </Link>
)

NextLink.propTypes = {
  href: PropTypes.object.isRequired,
  children: PropTypes.node.isRequired,
  as: PropTypes.string.isRequired
}

const capitalizeText = text =>
  text.split(' ').reduce((acc, val) => `${acc} ${capitalize(val)}`, '')

const delay = ms => new Promise(resolve => setTimeout(resolve, ms))

export const isImg = /^http(s)?:\/\/([\w-]+\.)+[\w-]+(\/[\w-./?%&=]*)?/
export const getChildrenToRender = (item, i) => {
  let tag = item.name.indexOf('title') === 0 ? 'h1' : 'div'
  tag = item.href ? 'a' : tag
  let children =
    typeof item.children === 'string' && item.children.match(isImg)
      ? React.createElement('img', { src: item.children, alt: 'img' })
      : item.children
  if (item.name.indexOf('button') === 0 && typeof item.children === 'object') {
    children = React.createElement(Button, {
      ...item.children
    })
  }
  return React.createElement(tag, { key: i.toString(), ...item }, children)
}

export { NextButtonLink, NextLink, isParent, isChild, capitalizeText, delay }
