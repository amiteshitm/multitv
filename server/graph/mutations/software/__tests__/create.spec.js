import { createTestClient } from 'apollo-server-testing'
import gql from 'graphql-tag'

import { constructTestServer } from 'test/apollo'
import { FactoryBot } from 'test/factories'

import { delay } from 'shared/util'

require('test/helper')
require('test/elastic')

const CREATE_SOFTWARE = gql`
  mutation SoftwareCreate($record: CreateOneSoftwareInput!) {
    softwareCreate(record: $record) {
      record {
        name
        slug
        details
        desc
        categories
        root_category {
          name
        }
      }
    }
  }
`

describe('mutation creates category', () => {
  let server, mutate, feature, category

  beforeEach(async () => {
    server = constructTestServer().server
    mutate = createTestClient(server).mutate

    feature = await FactoryBot.create('feature')
    category = await FactoryBot.create('category', {
      features: [feature._id],
    })
  })

  it('raises appropriate errors with invalid inputs', async () => {
    const res = await mutate({
      mutation: CREATE_SOFTWARE,
      variables: {
        record: {
          name: 'Test Software Invalid',
          slug: 'test-software',
          details: 'This is the first software',
          desc: 'This is a software description to test',
          root_category_id: category._id,
        },
      },
    })
    expect(res.errors).toHaveLength(1)
    expect(res.errors[0].name).toEqual('GraphQLError')
    expect(res.errors[0].extensions.code).toEqual('BAD_USER_INPUT')
  })

  describe('software creation', () => {
    beforeEach(async () => {
      jest.setTimeout(10000)
    })

    it('creates software with valid inputs', async () => {
      const res = await mutate({
        mutation: CREATE_SOFTWARE,
        variables: {
          record: {
            name: 'Test Software',
            slug: 'test-software',
            details: 'This is the first software',
            desc: 'This is a software description to test',
            root_category_id: category._id.toString(),
            categories: [category._id.toString()],
          },
        },
      })
      await delay(1000)
      expect(res.errors).toBeUndefined()
      expect(res.data.softwareCreate).toMatchSnapshot({
        record: {
          categories: [expect.any(String)],
          root_category: {
            name: expect.any(String),
          },
        },
      })
    })
  })
})
