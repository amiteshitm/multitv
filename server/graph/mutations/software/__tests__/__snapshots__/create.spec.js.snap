// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutation creates category software creation creates software with valid inputs 1`] = `
Object {
  "record": Object {
    "categories": Array [
      Any<String>,
    ],
    "desc": "This is a software description to test",
    "details": "This is the first software",
    "name": "Test Software",
    "root_category": Object {
      "name": Any<String>,
    },
    "slug": "test-software",
  },
}
`;
